File: HWs/TimeSeriesDeseasonalization/scripts/matlab/deseasonalization.m
data = readtable('../../data/avocado_exports.csv', 'VariableNamingRule', 'preserve');
df.Ano = double(data.("Año"));
df.Bimestre = double(data.Bimestre);
df.t = double(data.t);
df.Toneladas = double(data.Toneladas);

media = mean(df.Toneladas);
desv = std(df.Toneladas);
cv = (desv / media) * 100;
fprintf('[1] Coeficiente de variación: %.2f%%\n', cv);

fprintf('\n[2] Estadísticas descriptivas (Toneladas):\n');
fprintf('Media: %.2f\n', mean(df.Toneladas));
fprintf('Desv. Est.: %.2f\n', std(df.Toneladas));
fprintf('Mínimo: %.2f\n', min(df.Toneladas));
fprintf('Máximo: %.2f\n', max(df.Toneladas));
fprintf('Q1: %.2f\n', prctile(df.Toneladas, 25));
fprintf('Q2: %.2f\n', prctile(df.Toneladas, 50));
fprintf('Q3: %.2f\n', prctile(df.Toneladas, 75));

figure;
plot(df.t, df.Toneladas, 'o-');
title('Serie bimestral de exportaciones de aguacate');
xlabel('t (bimestres desde 2019-1)');
ylabel('Toneladas');
grid on;
saveas(gcf, '../../plots/matlab/avocado_exports_bimestral.png');
close;

figure;
histogram(df.Toneladas, 10);
title('Distribución de exportaciones (Toneladas)');
xlabel('Toneladas');
ylabel('Frecuencia');
grid on;
saveas(gcf, '../../plots/matlab/avocado_exports_histogram.png');
close;

figure;
boxplot(df.Toneladas, 'Labels', {'Toneladas'});
title('Boxplot de exportaciones bimestrales (Toneladas)');
ylabel('Toneladas');
grid on;
saveas(gcf, '../../plots/matlab/avocado_exports_boxplot.png');
close;

X = [ones(length(df.t), 1), df.t];
y = df.Toneladas;
[beta_orig, ~, ~, ~, stats_orig] = regress(y, X);

pendiente = beta_orig(2) / 1000.0;
fprintf('\n[3] Coeficiente de pendiente muestral (miles de toneladas por bimestre): %.4f\n', pendiente);

fprintf('\n[4] Prueba de hipótesis para pendiente > 0 (serie original)\n');
fprintf('R^2: %.4f\n', stats_orig(1));
fprintf('F-statistic: %.4f\n', stats_orig(2));
fprintf('p-value: %.4f\n', stats_orig(3));

n = length(y);
p = 2; % number of parameters
dof = n - p;
residuals = y - X * beta_orig;
mse = sum(residuals.^2) / dof;
se_beta = sqrt(mse * inv(X' * X));
t_stat = beta_orig(2) / se_beta(2, 2);
p_two_sided = 2 * (1 - tcdf(abs(t_stat), dof));
p_one_sided = p_two_sided / 2;
if t_stat < 0
    p_one_sided = 1 - p_one_sided;
end

fprintf('t = %.3f, p una-cola = %.4g -> %s\n', t_stat, p_one_sided, ...
    ternary(p_one_sided < 0.05, 'Rechazamos H0: tendencia positiva', 'No se rechaza H0'));

k = 6;
promedio_global = mean(df.Toneladas);
indices_estacionales = zeros(k, 1);
for bim = 1:k
    indices_estacionales(bim) = mean(df.Toneladas(df.Bimestre == bim)) / promedio_global;
end

fprintf('\n[5] Índices estacionales (multiplicativos):\n');
for bim = 1:k
    fprintf('Bimestre %d: %.4f\n', bim, indices_estacionales(bim));
end
fprintf('Interpretación: valores >1 indican bimestres por encima del promedio; <1 por debajo.\n');

df.IndiceEstacional = indices_estacionales(df.Bimestre);
df.Deseasonalizada = df.Toneladas ./ df.IndiceEstacional;

figure;
plot(df.t, df.Deseasonalizada, 'o-');
title('Serie desestacionalizada (multiplicativa)');
xlabel('t');
ylabel('Toneladas desestacionalizadas');
grid on;
saveas(gcf, '../../plots/matlab/avocado_exports_deseasonalized.png');
close;

X_d = [ones(length(df.t), 1), df.t];
y_d = df.Deseasonalizada;
[beta_des, ~, ~, ~, stats_des] = regress(y_d, X_d);

fprintf('\n[6] Regresión con datos desestacionalizados:\n');
fprintf('R^2: %.4f\n', stats_des(1));
fprintf('F-statistic: %.4f\n', stats_des(2));
fprintf('p-value: %.4f\n', stats_des(3));

residuals_d = y_d - X_d * beta_des;
mse_d = sum(residuals_d.^2) / dof;
se_beta_d = sqrt(mse_d * inv(X_d' * X_d));
t_stat_d = beta_des(2) / se_beta_d(2, 2);
p_two_sided_d = 2 * (1 - tcdf(abs(t_stat_d), dof));
p_one_sided_d = p_two_sided_d / 2;
if t_stat_d < 0
    p_one_sided_d = 1 - p_one_sided_d;
end

fprintf('t = %.3f, p una-cola = %.4g -> %s\n', t_stat_d, p_one_sided_d, ...
    ternary(p_one_sided_d < 0.05, 'Rechazamos H0: tendencia positiva', 'No se rechaza H0'));

t_future = 32;
X_future = [1, t_future];
pred_mean = X_future * beta_orig;

t_critical = tinv(0.975, dof);
se_pred = sqrt(mse * (X_future * inv(X' * X) * X_future'));
ci_low = pred_mean - t_critical * se_pred;
ci_high = pred_mean + t_critical * se_pred;

fprintf('\n[7] IC 95%% para la tendencia (media esperada) en 2024-bim2 (t=32): [%.2f, %.2f] toneladas. Estimado: %.2f\n', ...
    ci_low, ci_high, pred_mean);

r2_orig = stats_orig(1);
r2_des = stats_des(1);
fprintf('\n[8] R^2 serie original: %.4f\n', r2_orig);
fprintf('[8] R^2 serie desestacionalizada: %.4f\n', r2_des);
fprintf('Explicación: al quitar la variación estacional, el componente sistemático por tiempo puede capturar mejor la tendencia subyacente (o a veces menos, si la estacionalidad ya explicaba variación alineada con t). Un cambio en R^2 refleja cuánto peso de la variabilidad se atribuye a la estacionalidad vs. la tendencia.\n');

fut.t = (31:36)';
fut.Ano = zeros(length(fut.t), 1);
fut.Bimestre = zeros(length(fut.t), 1);

function [year, bim] = t_to_year_bim(t)
    base_year = 2019;
    idx = t - 1;
    year = base_year + floor(idx / 6);
    bim = mod(idx, 6) + 1;
end

for i = 1:length(fut.t)
    [fut.Ano(i), fut.Bimestre(i)] = t_to_year_bim(fut.t(i));
end

X_fut = [ones(length(fut.t), 1), fut.t];
pred_orig = X_fut * beta_orig;
fut.Pronostico_Original = pred_orig;

pred_des = X_fut * beta_des;
fut.IndiceEstacional = indices_estacionales(fut.Bimestre);
fut.Pronostico_Deseason = pred_des .* fut.IndiceEstacional;

fprintf('\n[9] Pronósticos 2024 (toneladas):\n');
fprintf('Año\tBimestre\tt\tPronóstico Original\tPronóstico Desestacionalizado\n');
for i = 1:length(fut.t)
    fprintf('%d\t%d\t\t%d\t%.2f\t\t\t%.2f\n', ...
        fut.Ano(i), fut.Bimestre(i), fut.t(i), ...
        fut.Pronostico_Original(i), fut.Pronostico_Deseason(i));
end

figure;
plot(df.t, df.Toneladas, 'o-', 'DisplayName', 'Observado');
hold on;
plot(fut.t, fut.Pronostico_Original, 'o--', 'DisplayName', 'Pronóstico (original)');
title('Pronóstico con modelo original');
xlabel('t');
ylabel('Toneladas');
legend('Location', 'best');
grid on;
saveas(gcf, '../../plots/matlab/avocado_exports_original_forecast.png');
close;

figure;
plot(df.t, df.Toneladas, 'o-', 'DisplayName', 'Observado');
hold on;
plot(fut.t, fut.Pronostico_Deseason, 'o--', 'DisplayName', 'Pronóstico (desestacionalizado)');
title('Pronóstico con modelo desestacionalizado (reestacionalizado)');
xlabel('t');
ylabel('Toneladas');
legend('Location', 'best');
grid on;
saveas(gcf, '../../plots/matlab/avocado_exports_deseasonalized_forecast.png');
close;

function result = ternary(condition, true_val, false_val)
    if condition
        result = true_val;
    else
        result = false_val;
    end
end

File: HWs/TimeSeriesDeseasonalization/scripts/r/deseasonalization.R
library(ggplot2)
library(dplyr)
library(gridExtra)

df <- read.csv("../../data/avocado_exports.csv")
df$Año <- as.integer(df$Año)
df$Bimestre <- as.integer(df$Bimestre)
df$t <- as.integer(df$t)
df$Toneladas <- as.numeric(df$Toneladas)

media <- mean(df$Toneladas)
desv <- sd(df$Toneladas)
cv <- (desv / media) * 100
cat(sprintf("[1] Coeficiente de variación: %.2f%%\n", cv))

cat("\n[2] Estadísticas descriptivas (Toneladas):\n")
print(summary(df$Toneladas))

p1 <- ggplot(df, aes(x = t, y = Toneladas)) +
  geom_line() +
  geom_point() +
  labs(title = "Serie bimestral de exportaciones de aguacate",
       x = "t (bimestres desde 2019-1)",
       y = "Toneladas") +
  theme_minimal() +
  theme(panel.grid = element_line(color = "gray90"))

ggsave("../../plots/r/avocado_exports_bimestral.png", p1, width = 8, height = 6)
print(p1)

p2 <- ggplot(df, aes(x = Toneladas)) +
  geom_histogram(bins = 10, fill = "steelblue", alpha = 0.7) +
  labs(title = "Distribución de exportaciones (Toneladas)",
       x = "Toneladas",
       y = "Frecuencia") +
  theme_minimal() +
  theme(panel.grid = element_line(color = "gray90"))

ggsave("../../plots/r/avocado_exports_histogram.png", p2, width = 8, height = 6)
print(p2)

p3 <- ggplot(df, aes(y = Toneladas)) +
  geom_boxplot(fill = "steelblue", alpha = 0.7) +
  labs(title = "Boxplot de exportaciones bimestrales (Toneladas)",
       y = "Toneladas") +
  theme_minimal() +
  theme(panel.grid = element_line(color = "gray90"))

ggsave("../../plots/r/avocado_exports_boxplot.png", p3, width = 8, height = 6)
print(p3)

model_orig <- lm(Toneladas ~ t, data = df)

pendiente <- coef(model_orig)["t"] / 1000.0
cat(sprintf("\n[3] Coeficiente de pendiente muestral (miles de toneladas por bimestre): %.4f\n", pendiente))

cat("\n[4] Prueba de hipótesis para pendiente > 0 (serie original)\n")
print(summary(model_orig))

t_stat <- summary(model_orig)$coefficients["t", "t value"]
p_two_sided <- summary(model_orig)$coefficients["t", "Pr(>|t|)"]
p_one_sided <- p_two_sided / 2
if (t_stat < 0) {
  p_one_sided <- 1 - p_one_sided
}

cat(sprintf("t = %.3f, p una-cola = %.4g -> %s\n", t_stat, p_one_sided,
            ifelse(p_one_sided < 0.05, "Rechazamos H0: tendencia positiva", "No se rechaza H0")))

k <- 6
promedio_global <- mean(df$Toneladas)
indices_estacionales <- numeric(k)
for (bim in 1:k) {
  indices_estacionales[bim] <- mean(df$Toneladas[df$Bimestre == bim]) / promedio_global
}

cat("\n[5] Índices estacionales (multiplicativos):\n")
for (bim in 1:k) {
  cat(sprintf("Bimestre %d: %.4f\n", bim, indices_estacionales[bim]))
}
cat("Interpretación: valores >1 indican bimestres por encima del promedio; <1 por debajo.\n")

df$IndiceEstacional <- indices_estacionales[df$Bimestre]
df$Deseasonalizada <- df$Toneladas / df$IndiceEstacional

p4 <- ggplot(df, aes(x = t, y = Deseasonalizada)) +
  geom_line() +
  geom_point() +
  labs(title = "Serie desestacionalizada (multiplicativa)",
       x = "t",
       y = "Toneladas desestacionalizadas") +
  theme_minimal() +
  theme(panel.grid = element_line(color = "gray90"))

ggsave("../../plots/r/avocado_exports_deseasonalized.png", p4, width = 8, height = 6)
print(p4)

model_des <- lm(Deseasonalizada ~ t, data = df)

cat("\n[6] Regresión con datos desestacionalizados:\n")
print(summary(model_des))

t_stat_d <- summary(model_des)$coefficients["t", "t value"]
p_two_sided_d <- summary(model_des)$coefficients["t", "Pr(>|t|)"]
p_one_sided_d <- p_two_sided_d / 2
if (t_stat_d < 0) {
  p_one_sided_d <- 1 - p_one_sided_d
}

cat(sprintf("t = %.3f, p una-cola = %.4g -> %s\n", t_stat_d, p_one_sided_d,
            ifelse(p_one_sided_d < 0.05, "Rechazamos H0: tendencia positiva", "No se rechaza H0")))

t_future <- 32
X_future <- data.frame(t = t_future)
pred_mean <- predict(model_orig, newdata = X_future, interval = "confidence", level = 0.95)
ci_low <- pred_mean[1, "lwr"]
ci_high <- pred_mean[1, "upr"]
mean_pred <- pred_mean[1, "fit"]

cat(sprintf("\n[7] IC 95%% para la tendencia (media esperada) en 2024-bim2 (t=32): [%.2f, %.2f] toneladas. Estimado: %.2f\n",
            ci_low, ci_high, mean_pred))

pi <- predict(model_orig, newdata = X_future, interval = "prediction", level = 0.95)
cat(sprintf("[7-extra] IC 95%% de predicción (valor observado esperado) en 2024-bim2: [%.2f, %.2f]\n",
            pi[1, "lwr"], pi[1, "upr"]))

r2_orig <- summary(model_orig)$r.squared
r2_des <- summary(model_des)$r.squared
cat(sprintf("\n[8] R^2 serie original: %.4f\n", r2_orig))
cat(sprintf("[8] R^2 serie desestacionalizada: %.4f\n", r2_des))
cat("Explicación: al quitar la variación estacional, el componente sistemático por tiempo puede capturar mejor la tendencia subyacente (o a veces menos, si la estacionalidad ya explicaba variación alineada con t). Un cambio en R^2 refleja cuánto peso de la variabilidad se atribuye a la estacionalidad vs. la tendencia.\n")

fut <- data.frame(t = 31:36)

t_to_year_bim <- function(t) {
  base_year <- 2019
  idx <- t - 1
  year <- base_year + floor(idx / 6)
  bim <- (idx %% 6) + 1
  return(c(year, bim))
}

yb <- t(sapply(fut$t, t_to_year_bim))
fut$Año <- yb[, 1]
fut$Bimestre <- yb[, 2]

pred_orig <- predict(model_orig, newdata = fut)
fut$Pronostico_Original <- pred_orig

pred_des <- predict(model_des, newdata = fut)
fut$IndiceEstacional <- indices_estacionales[fut$Bimestre]
fut$Pronostico_Deseason <- pred_des * fut$IndiceEstacional

cat("\n[9] Pronósticos 2024 (toneladas):\n")
print(round(fut[, c("Año", "Bimestre", "t", "Pronostico_Original", "Pronostico_Deseason")], 2))

pred_orig_full <- predict(model_orig, newdata = fut, interval = "confidence", level = 0.95)
fut$IC_orig_low <- pred_orig_full[, "lwr"]
fut$IC_orig_high <- pred_orig_full[, "upr"]

pred_des_full <- predict(model_des, newdata = fut, interval = "prediction", level = 0.95)
fut$IC_des_low <- pred_des_full[, "lwr"] * fut$IndiceEstacional
fut$IC_des_high <- pred_des_full[, "upr"] * fut$IndiceEstacional

cat("\n[9-extra] Pronósticos con intervalos de confianza:\n")
print(round(fut[, c("Año", "Bimestre", "Pronostico_Original", "IC_orig_low", "IC_orig_high",
                     "Pronostico_Deseason", "IC_des_low", "IC_des_high")], 2))

p5 <- ggplot() +
  geom_line(data = df, aes(x = t, y = Toneladas), color = "black") +
  geom_point(data = df, aes(x = t, y = Toneladas), color = "black") +
  geom_line(data = fut, aes(x = t, y = Pronostico_Original), color = "red", linetype = "dashed") +
  geom_point(data = fut, aes(x = t, y = Pronostico_Original), color = "red") +
  labs(title = "Pronóstico con modelo original",
       x = "t",
       y = "Toneladas") +
  theme_minimal() +
  theme(panel.grid = element_line(color = "gray90"))

ggsave("../../plots/r/avocado_exports_original_forecast.png", p5, width = 8, height = 6)
print(p5)

p6 <- ggplot() +
  geom_line(data = df, aes(x = t, y = Toneladas), color = "black") +
  geom_point(data = df, aes(x = t, y = Toneladas), color = "black") +
  geom_line(data = fut, aes(x = t, y = Pronostico_Deseason), color = "blue", linetype = "dashed") +
  geom_point(data = fut, aes(x = t, y = Pronostico_Deseason), color = "blue") +
  labs(title = "Pronóstico con modelo desestacionalizado (reestacionalizado)",
       x = "t",
       y = "Toneladas") +
  theme_minimal() +
  theme(panel.grid = element_line(color = "gray90"))

ggsave("../../plots/r/avocado_exports_deseasonalized_forecast.png", p6, width = 8, height = 6)
print(p6)
